content/software-testing-the-hardware-software-interface-part-2.md

```diff
@@ -0,0 +1,417 @@
+previous testing code a bit to use that telnet connection instead of the local UART:
+
+```python
+import unittest
+import telnetlib3
+import asyncio
+import time
+import xmlrunner
+
+
+class TelnetPromptTest(unittest.TestCase):
+    TELNET_HOST = "192.168.15.3"
+    TELNET_PORT = 9000
+    TIMEOUT = 40
+    PROMPT = "verdin-imx8mp-06817296 login:"
+
+    async def check_prompt(self):
+        reader, writer = await telnetlib3.open_connection(self.TELNET_HOST, self.TELNET_PORT)
+        start_time = time.time()
+        received_data = ""
+
+        try:
+            while time.time() - start_time < self.TIMEOUT:
+                chunk = await reader.read(1024)
+                if chunk:
+                    print(chunk, end="")
+                    received_data += chunk
+
+                    if self.PROMPT in received_data:
+                        print(f"\nFound expected prompt '{self.PROMPT}'")
+                        return True
+                else:
+                    await asyncio.sleep(0.1)
+
+            self.fail(f"Expected prompt '{self.PROMPT}' not found within {self.TIMEOUT} seconds.")
+        finally:
+            writer.close()
+
+    def test_wait_for_PROMPT(self):
+        asyncio.run(self.check_prompt())
+
+
+if __name__ == "__main__":
+    with open("telnet_test_results.xml", "wb") as output:
+        unittest.main(testRunner=xmlrunner.XMLTestRunner(output=output), exit=False)
+```
+
+which yields the same results as before.
+
+Because the script can run anywhere now, integrating everything within a CI/CD systems should be fairly straight forward.
+Let's do it with Jenkins, which is a popular and mature CI/CD system. For that, I'll write a simple `Jenkinsfile` that
+will execute the tests and parse the results for us:
+
+```groovy
+pipeline {
+    agent any
+
+    stages {
+        stage('Setup') {
+            steps {
+                echo 'Setting up Python environment'
+                sh 'python3 -m venv venv'
+                sh '. venv/bin/activate && pip install -r requirements.txt'
+            }
+        }
+
+        stage('Run Telnet Test') {
+            steps {
+                echo 'Running Telnet Test'
+                sh '. venv/bin/activate && python3 boot-test-telnet.py'
+            }
+        }
+    }
+
+    post {
+        always {
+            echo 'Cleaning up virtual environment'
+            sh 'rm -rf venv'
+        }
+
+        success {
+            echo 'Tests ran successfully'
+        }
+
+        failure {
+            echo 'Tests failed, please check the results.'
+        }
+
+        unstable {
+            junit 'telnet_test_results.xml'
+        }
+    }
+}
+```
+
+Which results in this nice CI job that is testing if our Verdin iMX8MP is properly booting! Cool, right?
+
+{{< figure src="/jenkins-result.png" title="Jenkins Pipeline Stages" alt="" width="100%" class="centered-figure">}}
```

Comment:
in fact all the images after rendered are pretty small, and is not possible to enlarge then on click.

Reply:
Tried to fix here, https://github.com/torizon/blog/pull/18#discussion_r1827670433. So closing this to keep the discussion in the first post.